### PR TITLE
fix: enhance FocusDirective with cxRefreshFocusOn input

### LINKED ISSUE #9531: Focus newly opened section in the b2b table component
### Current
While tabbing through the b2b table component, opening  up a new section (on the right) is not making the focus to go to the newly opened section, but rather the focus stays where it was (where we pressed enter).

### Expected
After opening up a new section, the focus should go to the first element of the section.

### PR DESCRIPTION
**Pull request by wangzixi-diablo**
closes: #9531

### CODE REVIEW

**Comment by wangzixi-diablo** [comment]
As a draft PR, currently only focusable element in Unit detail page (unit-details.component.html) is automatically focused. Wait for expert's feedback and then continue.

**Comment by wangzixi-diablo** [comment]
As a draft PR, currently only focusable element in Unit detail page (unit-details.component.html) is automatically focused. Wait for expert's feedback and then continue.

# how to test

Choose one table row in Unit list page:

image

image

When the detail page is already rendered, change the selection of unit, and detail page will be refreshed with new unit data loaded from backend. After detail page is refreshed, again Edit element will be focused automatically:

image

# drawback of this enhancement

Take this example for illustration: when the detail page is rendered for the first time, Edit element will be focused by code **TWICE**:

**Firstly**, it's focused because of the following code in focus.directive.ts:

```typescript
@Input() set cxRefreshFocusOn(_switchCondition: string) {
    if (this.shouldAutofocus) {
      super.handleFocus(); -------------------- this line focuses the Edit element
    }
  }
```

**Secondly**, it's focused because of the following code in auto-focus.directive.ts:

```typescript
ngAfterViewInit(): void {
    if (this.shouldAutofocus) {
      this.handleFocus(); -------------------- this line focuses the Edit element
    }
    if (!this.shouldAutofocus || this.hasPersistedFocus) {
      super.ngAfterViewInit();
    }
  }
```

After unit detail view is rendered, its ngAfterViewInit will not be triggered, thus no duplicated focus will occur.

File: feature-libs/organization/administration/components/unit/details/unit-details.component.html
Comment (tobi-or-not-tobi) [inline feature-libs/organization/administration/components/unit/details/unit-details.component.html]: I'd not introduce a new DOM element here, you can apply the condition and config directly to the anchor.

File: feature-libs/organization/administration/components/unit/details/unit-details.component.html
Comment (tobi-or-not-tobi) [inline feature-libs/organization/administration/components/unit/details/unit-details.component.html]: The cxFocus comes with a configuration. I'd consider extending the `AutoFocusConfig` with a `refreshFocus` property.

Comment (tobi-or-not-tobi) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: worth trying if that gives the right result though.

**Review by tobi-or-not-tobi — COMMENTED** [review]
This is a great start. During testing, I noticed that the edit button gets a -1 tabindex. This means that after leaving the element and tabbing back, you can't select it anymore. Something is off.

File: feature-libs/organization/administration/components/unit/details/unit-details.component.html
Comment (wangzixi-diablo) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: Hi Tobias, 

Thanks a lot for your kind review! Let me explain why I introduce a wrapper div element on purpose.

Yes, I also observed that the host element which declares [cxFocus] directive will have tabindex set as -1. 

I debugged and knew the reason, tabindex is set as -1 by default in base-focus.directive.ts:

image

In order to avoid this issue, I introduce a wrapper div element, and declare [cxFocus] directive on it. In this way, div itself has tabindex = -1, while edit button, the a tag, is still focusable:

image

So do you know if there's any approach, to avoid the "side-effect" of having host element's tabindex attribute set as default value -1? I do believe there's a way, otherwise it does not make sense. 

lock+Focus (1)

I tried to find some hint in lock-focus.directive.ts but unfortunately didn't find what I want - in my eye its logic is complex, I draw a mindmap to try to understand it...

Thanks a lot for your help!

Comment (tobi-or-not-tobi) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: understood. I'd still like to avoid the div, can you debut on why `requiresExplicitTabIndex` is returning `true` in this case?

Comment (wangzixi-diablo) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: Hi Tobias, 

I am fine with removal of this unnecessary div wrapper. Now <a> tag looks like below. I didn't add refresh function since it is not related to "tabindex = -1" trouble shooting.

```html
<a
      actions *ngIf="model.active"
      class="link edit"
      [class.disabled]="!model.active || (isInEditMode$ | async)"
      [routerLink]="{ cxRoute: 'orgUnitEdit', params: model } | cxUrl"
      [cxFocus]="{ autofocus: true }"
    >
      {{ 'organization.edit' | cxTranslate }}
    </a>
```

when debugging into function **requiresExplicitTabIndex**, it returns true because evaluation '**(this.host.hasAttribute('href') || this.host.hasAttribute('routerlink'))**' returns false.

image

In the context of **requiresExplicitTabIndex** execution, <a> tag does not have href attribute populated yet. 

Neither does it has "routerlink", since at this time, "routerlink" is already converted to **ng-reflect-router-link** by Angular.

image

And as far as I know, href attribute of <a> tag only gets generated when another round of Angular change detection is performed. Unfortunately at this time the evaluation of [cxFocus] is already done, no chance to re-evaluated...

image

# My Summary

I would describe this behavior as that [cxFocus] has a time dependency with [routerlink], what do you think? 

Of course, if I replace 'routerlink' with 'ng-reflect-router-link' in below code, the whole scenario will work, no tabindex=-1 any more. However since ng-reflect-router-link is internally populated by Angular, I don't think it's a practice to introduce it in our own code. 

image

**Comment by wangzixi-diablo** [comment]
1. move this.requiredtabindex = -1 into ngAfterViewInit, due to two reasons:

a. before change: In the context of requiresExplicitTabIndex execution, tag does not have href attribute populated yet. For detail see this discussion [thread](https://github.com/SAP/spartacus/pull/10507#discussion_r553936239)

as a result, we have to put it to ngOnAfterViewInit, at that time the href attribute of a tag is already populated with correct value：

clipboard1


b. The problem here is that we have a situation where the view generation process (which ngAfterViewInit is a part of) is itself further modifying the data that we are trying to display in the first place.

In unit test code it reports error message:

> ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'undefined'. Current value: '-1'.

See this [blog](https://blog.angular-university.io/angular-debugging/).

So I use setTimeout() to defer the code to another Javascript Virtual Machine turn.

2. Since now this.requiredtabindex = -1 is wrapped to setTimeout and executed asynchronously, some unit test will fail:

clipboard2



3. how to use this extended attribute in units-detail.component.html?

clipboard3

File: projects/storefrontlib/src/layout/a11y/keyboard-focus/base/base-focus.directive.ts
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/layout/a11y/keyboard-focus/base/base-focus.directive.ts]: we don't like this in our code. There are other tricks to make unit tests work. You might want to override the change detection strategy in unit tests.

File: projects/storefrontlib/src/layout/a11y/keyboard-focus/focus.directive.ts
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/layout/a11y/keyboard-focus/focus.directive.ts]: I'd not introduce this new input, but rather introduce another property in the `FocusConfig`

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

**Comment by wangzixi-diablo** [comment]
# Some background of my latest commit ( 2021-2-2)

1. handleFocus method is called twice in different hooks in different focus class, which have different focus:

* in focus.directive.ts, ngOnChanges:

this hook is responsible to refresh focus whenever component property configured in focus configuration ( config.refreshFocus ) is changed.

* in auto-focus.directive.ts, ngOnAfterViewInit:

this hook is called when the view is fully rendered, responsible to auto-focus an element with autofocus = true.

In real scenario, for example when we opened unit list in B2B, click one unit to reach unit detail page(unit-details.component.html), an element marked with [cxFocus]="{ autofocus: true, refreshFocus: model.uid }" will be focused TWICE, one in ngOnChanges and the other one in ngAfterViewInit hook.

2. I didn't implement OnChanges in base-focus.directive.ts, but in focus.directive.ts instead.

The reason is, I checked in base-focus.directive.ts, the ngOnInit hook implemented there has some default logic. So if I implement OnChanges hook in base-focus.directive.ts, I don't have any default logic to code there, thus will lead to warnings in Visual Studio Code. So I choose to implement OnChanges in focus.directive.ts instead, this is current design in my PR.

File: feature-libs/organization/administration/components/unit/details/unit-details.component.html
Comment (tobi-or-not-tobi) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: Good summary, and i'm not against using that angular internal to be frank. We could do a better job, but it takes a refactor that we cannot easily do without breaking changes. I've used the ng internal for now to unblock us with this behaviour. It's hitting us in more locations, so i'm eager to get this fixed.

File: feature-libs/organization/administration/components/unit/details/unit-details.component.html
Comment (Pucek9) [inline feature-libs/organization/administration/components/unit/details/unit-details.component.html]: Is this change necessary? We are using already class=disabled for that unable to open edit form.
I tnink it can be breaking change and doesnt look good for mobile devices
```suggestion
```

Comment (tobi-or-not-tobi) [reply feature-libs/organization/administration/components/unit/details/unit-details.component.html]: probably not actually. This was apparently added previously to mitigate a behaviour that I fixed in the focus directive. I wasn't aware of the ngif being added.

**Review by tobi-or-not-tobi — APPROVED** [review]

**Pull request merged**
